nixos/tests/k3s/etcd.nix

```diff
@@ -0,0 +1,101 @@
+import ../make-test-python.nix ({ pkgs, lib, k3s, ... }:
+
+{
+  name = "${k3s.name}-etcd";
+
+  nodes = {
+
+    etcd = { ... }: {
+      services.etcd = {
+        enable = true;
+        openFirewall = true;
+        listenClientUrls = [ "http://192.168.1.1:2379" "http://127.0.0.1:2379" ];
+        listenPeerUrls = [ "http://192.168.1.1:2380" ];
+        initialAdvertisePeerUrls = [ "http://192.168.1.1:2380" ];
+        initialCluster = [ "etcd=http://192.168.1.1:2380" ];
+      };
+      networking = {
+        useDHCP = false;
+        defaultGateway = "192.168.1.1";
+        interfaces.eth1.ipv4.addresses = pkgs.lib.mkForce [
+          { address = "192.168.1.1"; prefixLength = 24; }
+        ];
+      };
+    };
+
+    k3s = { pkgs, ... }: {
+      environment.systemPackages = with pkgs; [ jq ];
+      # k3s uses enough resources the default vm fails.
+      virtualisation.memorySize = 1536;
+      virtualisation.diskSize = 4096;
+
+      services.k3s = {
+        enable = true;
+        role = "server";
+        extraFlags = builtins.toString [
+          "--datastore-endpoint=\"http://192.168.1.1:2379\""
+          "--disable" "coredns"
+          "--disable" "local-storage"
+          "--disable" "metrics-server"
+          "--disable" "servicelb"
+          "--disable" "traefik"
+          "--node-ip" "192.168.1.2"
+          "--pause-image" "test.local/pause:local"
+        ];
+      };
+
+      networking = {
+        firewall = {
+          allowedTCPPorts = [ 2379 2380 6443 ];
+          allowedUDPPorts = [ 8472 ];
+        };
+        useDHCP = false;
+        defaultGateway = "192.168.1.2";
+        interfaces.eth1.ipv4.addresses = pkgs.lib.mkForce [
+          { address = "192.168.1.2"; prefixLength = 24; }
+        ];
+      };
+    };
+
+  };
+
+  testScript = ''
+    with subtest("should start etcd"):
+        etcd.start()
+        etcd.wait_for_unit("etcd.service")
+
+    with subtest("should wait for etcdctl endpoint status to succeed"):
+        etcd.wait_until_succeeds("etcdctl endpoint status")
+
+    with subtest("should start k3s"):
+        k3s.start()
+        k3s.wait_for_unit("k3s")
+
+    with subtest("should test if kubectl works"):
+        k3s.wait_until_succeeds("k3s kubectl get node")
+
+    with subtest("should wait for service account to show up; takes a sec"):
+        k3s.wait_until_succeeds("k3s kubectl get serviceaccount default")
+
+    with subtest("should create a sample secret object"):
+        k3s.succeed("k3s kubectl create secret generic nixossecret --from-literal thesecret=abacadabra")
+
+    with subtest("should check if secret is correct"):
+        k3s.wait_until_succeeds("test $(kubectl get secrets nixossecret -o json | jq -r .data.thesecret | base64 -d) == abacadabra")
+
+    with subtest("should have a secret in database"):
+        etcd.wait_until_succeeds("[[ $(etcdctl get /registry/secrets/default/nixossecret | head -c1 | wc -c) -ne 0 ]]")
```

Comment:
This feels a little odd to test.

I think the etcd structure k8s uses isn't guaranteed and can be broken, and for example has been in the past with the etcd 2-> 3 migration.

Because of that, it feels a little fragile to read this structure out of etcd.

I do get the appeal of verifying it's using the etcd backend, and I don't think they're going to change the structure anytime soon (it feels like it might be ossified at this point), but it still feels like a wrong thing to test.

I'm not going to block on this, just observe that I think k8s is within its right to break this test.

Reply:
As an integration test, both sides need to be sanity checked.
Whenever the resource path changes, we have to update it. But then, we can trust that k3s is actually using etcd (and not sqlite). Otherwise, I don't know how to test if etcd is actually being used.
I'm open to alternative implementations.

Reply:
> This feels a little odd to test.
> k8s is within its right to break this test.

A thought to consider: A test automates our debug process. Whenever something (`k8s`) breaks our debug process, we should update our debug process. So that breakage is welcomed because our debug process is stale.

I see a beauty in this. :-)

Reply:
The thing we really want confidence for here is that the nixos module / k3s configured to use etcd are actually using etcd, and the exposed user functionality works.

If this was an integration test for k3s, testing that etcd's contents are what's expected would make sense since that's part of k3s's integration with etcd.
In NixOS's case, the surface area we care about is just that the k3s config options make k3s actually use etcd, and the functionality of k3s works.

This does test that k3s is actually using etcd well, so in that sense it's a good test for sure! And in practice, I don't know of another way to test that, so this might be our best bet.

I didn't realize it tested something you actually used yourself as well, which is another point in factor of the test being useful.

Anyway, taking it all together, I'm fine merging this test. In practice I expect it to be stable, and if it turns out k8s does change the etcd structure more often than, say, once every 2 years, we can reconsider the test then.

Thanks for writing it!

Reply:
I get that you are **not** in opposition to this. But just wanted to clarify a nuance. Just so we are at same page. I think we should eventually agree on this. Because it is a logical conclusion. And likely I am failing to communicate it properly.

> If this **was** an integration test for k3s

I think this is still an integration test. The nuance is:
Upstream k3s tests their k3s with their etcd.
Here is being tested nixpkgs's k3s with nixpkgs's etcd (using NixOS configuration module).

> if ... k8s ... change the etcd structure ... we can **reconsider the test** then.

If this test ever gets removed, it's not possible to guarantee etcd side is working as expected.
K3s could be using it's embedded sqlite and we would never know.
On breakage, I think, the test should be updated to reflect the new debug process.

Reply:
Yup, understood. I think we agree overall.

Being sure that k3s is using etcd is a critical part of the test, and this is the mechanism we've picked for that.

Thanks for talking through it more!